### PR TITLE
feat: allow composing data with abi

### PR DESCRIPTION
**Pull request by antoncoding**
### TLDR
* Allow user to generate `data` input by choosing function + parameters from `to` address
* We use another modal to abstract away this logic, once generated, the modal will be closed + data will filled

### Testing
Can test with USDC transfer with the following addresses:
* [USDC Faucet](https://faucet.circle.com/)
* [USDC contract ](https://sepolia-optimism.etherscan.io/address/0x5fd84259d66cd46123540766be93dfe6d43130d7#readProxyContract)
* [USDC implementation](https://sepolia-optimism.etherscan.io/address/0x101e1c9757c99867a39ceb41117b44f2292cb253#writeContract)

### Misc
the following 2 env variables are **not required** on testnet, only required when we deploy for mainnet

```
NEXT_PUBLIC_OPTIMISM_ETHERSCAN_API_KEY
NEXT_PUBLIC_ARBITRUM_ETHERSCAN_API_KEY
```

### CODE REVIEW

File: src/components/SmartModeInput/index.tsx [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+    } catch (error) {
+      console.error('Error encoding function data:', error);
+      toast.error('Invalid input data');
+    }
+  };
+
+  const writableFunctions =
+    abi?.filter(
+      (item: any) =>
+        item.type === 'function' && !item.stateMutability.includes('view')
```
Comment (NIC619) [inline src/components/SmartModeInput/index.tsx]: Do we also need to exclude `pure`?

File: src/components/SmartModeInput/index.tsx [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+      <Tooltip content="Use this if the contract is a proxy. Enter the implementation contract address.">
+        <Input
+          label="Proxy Implementation Address (optional)"
+          placeholder="0x"
+          type="text"
+          value={proxyImplementationAddress}
+          onChange={(e) => setProxyImplementationAddress(e.target.value)}
+          size="sm"
+        />
+      </Tooltip>
```
Comment (NIC619) [inline src/components/SmartModeInput/index.tsx]: I think we can leverage etherscan's api to automatically pulled the implementation address of a verified proxy contract on etherscan: https://ethereum.stackexchange.com/questions/135333/is-it-possible-to-pull-proxy-contracts-through-an-ethereum-based-api

This is the result queried with `https://api-sepolia-optimistic.etherscan.io/api?module=contract&action=getsourcecode&address=0x5fd84259d66Cd46123540766Be93DFE6D43130D7`:
```
{
  "status": "1",
  "message": "OK",
  "result": [
    {
      "SourceCode": ...
      ...
      "Implementation": "0x101e1c9757c99867a39ceb41117b44f2292cb253",
      ...
    }
  ]
}
```

Comment (NIC619) [reply src/components/SmartModeInput/index.tsx]: And if the proxy is not verified on etherscan, then we provide this additional manual input tool in case user needs it.

File: src/components/SmartModeInput/index.tsx (hunk 1 above)
Comment (antoncoding) [reply src/components/SmartModeInput/index.tsx]: good catch!

**Comment by antoncoding** [comment]
with the new change:
* automatically detect proxy contract
* allow changing the proxy contract, or allow manually setting if no proxy is detected

Screebsgit: proxy detected / not detected
Screenshot 2024-10-07 at 11 41 13Screenshot 2024-10-07 at 11 43 10

**Review by NIC619 — APPROVED** [review]
